pkg/build/controller/build/build_controller.go

```diff
@@ -372,14 +383,24 @@ func (bc *BuildController) createPodSpec(originalBuild *buildapi.Build, ref stri
 	return podSpec, nil
 }
 
-// resolveOutputDockerImageReference returns a reference to a Docker image
-// computed from the buid.Spec.Output.To reference.
-func (bc *BuildController) resolveOutputDockerImageReference(build *buildapi.Build) (string, error) {
+// resolveOutputDockerImageReference returns a Docker image name
+// computed from the build.Spec.Output.To reference.  It also attempts to find a Secret
+// associated with the build service account which can push to the image location and
+// return that, if found.
+// Note that we are using controller level permissions to resolve the imagestream and secret,
+// meaning users could theoretically define a build that references an imagestream they cannot
+// see, and 1) get the docker image reference of that imagestream and 2) a reference to a secret
+// associated with a service account that can push to that location.  However they still cannot view the secret,
+// and ability to use a service account implies access to its secrets, so this is considered safe.
+// Furthermore it's necessary to enable triggered builds since a triggered build is not "requested"
+// by a particular user, so there are no user permissions to validate against in that case.
+func (bc *BuildController) resolveOutputDockerImageReference(build *buildapi.Build) (string, *kapi.LocalObjectReference, error) {
 	outputTo := build.Spec.Output.To
 	if outputTo == nil || outputTo.Name == "" {
-		return "", nil
+		return "", build.Spec.Output.PushSecret, nil
```

Comment:
fair enough.  I guess I'm looking (in vain? ;) for ways to make this function easier for people to reason about.  The alternative which I suspect you've considered already is to split it into two separate functions - resolve the reference, and find the secret.  I'm prepared for that suggestion not to stick as well ;)